File: kits/data-integration/providers/capture_mode/ShareIntent.swift
// Data Integration Kit - Share Intent Capture Provider
// Mobile/OS share sheet receiver normalizing across iOS NSItemProvider and Android Intent extras

import Foundation

public enum SharePlatform: String {
    case ios, android, web, desktop, unknown
}

public enum ShareContentType: String {
    case text, url, image, file, mixed

    var mimeType: String {
        switch self {
        case .image: return "image/*"
        case .url: return "text/uri-list"
        case .file: return "application/octet-stream"
        case .text, .mixed: return "text/plain"
        }
    }
}

public struct SharedImage {
    public var mimeType: String
    public var width: Int?
    public var height: Int?
}

public struct SharedFile {
    public var mimeType: String
    public var filename: String
    public var size: Int
}

public struct NormalizedShareData {
    public var text: String?
    public var url: String?
    public var title: String?
    public var images: [SharedImage] = []
    public var files: [SharedFile] = []
    public var platform: SharePlatform
    public var contentType: ShareContentType
}

public final class ShareIntentCaptureProvider {
    public init() {}

    public func capture(input: CaptureInput, config: CaptureConfig) async throws -> CaptureItem {
        guard let shareData = input.shareData else {
            throw CaptureError.parseError("share_intent capture requires shareData")
        }

        let normalized = normalizeShareData(shareData)

        var contentParts: [String] = []
        if let title = normalized.title { contentParts.append("# \(title)") }
        if let url = normalized.url { contentParts.append("URL: \(url)") }
        if let text = normalized.text { contentParts.append(text) }
        if !normalized.images.isEmpty {
            contentParts.append("\nImages: \(normalized.images.count) shared")
            for (i, img) in normalized.images.enumerated() {
                let dimInfo = img.width.map { " \($0)x\(img.height ?? 0)" } ?? ""
                contentParts.append("  [\(i + 1)] \(img.mimeType)\(dimInfo)")
            }
        }
        if !normalized.files.isEmpty {
            contentParts.append("\nFiles: \(normalized.files.count) shared")
            for (i, file) in normalized.files.enumerated() {
                contentParts.append("  [\(i + 1)] \(file.filename) (\(file.mimeType), \(file.size) bytes)")
            }
        }

        let title = normalized.title
            ?? normalized.url.flatMap { URL(string: $0)?.host }.map { "Shared: \($0)" }
            ?? "Shared Content"

        let content = contentParts.isEmpty ? "(empty share)" : contentParts.joined(separator: "\n")

        var tags = ["share-intent", normalized.platform.rawValue, normalized.contentType.rawValue]
        if !normalized.images.isEmpty { tags.append("has-images") }
        if !normalized.files.isEmpty { tags.append("has-files") }

        return CaptureItem(
            content: content,
            sourceMetadata: SourceMetadata(
                title: title,
                url: normalized.url,
                capturedAt: ISO8601DateFormatter().string(from: Date()),
                contentType: normalized.contentType.mimeType,
                author: nil,
                tags: tags,
                source: "share_intent"
            ),
            rawData: (config.options?["includeRaw"] as? Bool == true) ? shareData : nil
        )
    }

    public func supports(input: CaptureInput) -> Bool {
        return input.shareData != nil
    }

    // MARK: - Platform Detection

    private func detectPlatform(_ data: Any) -> SharePlatform {
        guard let dict = data as? [String: Any] else { return .unknown }

        // iOS NSItemProvider pattern
        if dict["itemProviders"] != nil || dict["NSExtensionItem"] != nil || dict["UTType"] != nil {
            return .ios
        }
        // Android Intent pattern
        if dict["action"] != nil || dict["EXTRA_TEXT"] != nil || dict["EXTRA_STREAM"] != nil {
            return .android
        }
        // Web Share API
        if dict["title"] != nil && (dict["text"] != nil || dict["url"] != nil) {
            if dict["action"] == nil && dict["itemProviders"] == nil { return .web }
        }
        // Desktop clipboard
        if dict["clipboardData"] != nil || dict["dataTransfer"] != nil {
            return .desktop
        }
        return .unknown
    }

    private func isUrl(_ s: String) -> Bool {
        s.hasPrefix("http://") || s.hasPrefix("https://")
    }

    // MARK: - Platform-Specific Normalization

    private func normalizeShareData(_ shareData: Any) -> NormalizedShareData {
        let platform = detectPlatform(shareData)
        switch platform {
        case .ios: return normalizeIos(shareData)
        case .android: return normalizeAndroid(shareData)
        case .web: return normalizeWeb(shareData)
        default: return normalizeGeneric(shareData, platform: platform)
        }
    }

    private func normalizeIos(_ data: Any) -> NormalizedShareData {
        guard let dict = data as? [String: Any] else {
            return NormalizedShareData(platform: .ios, contentType: .text)
        }

        var result = NormalizedShareData(platform: .ios, contentType: .text)

        // NSExtensionItem title
        if let ext = dict["NSExtensionItem"] as? [String: Any],
           let title = ext["attributedTitle"] as? String {
            result.title = title
        }

        // NSItemProvider items
        let items = (dict["itemProviders"] ?? dict["items"]) as? [[String: Any]] ?? []
        for item in items {
            let uti = (item["UTType"] ?? item["typeIdentifier"]) as? String ?? ""
            let dataStr = item["data"] as? String ?? ""

            if uti.contains("public.url") || uti.contains("public.plain-text") {
                if isUrl(dataStr) {
                    result.url = dataStr
                    result.contentType = .url
                } else {
                    result.text = dataStr
                }
            } else if uti.contains("public.image") {
                let mime = uti.contains("png") ? "image/png" : "image/jpeg"
                result.images.append(SharedImage(mimeType: mime))
                result.contentType = (result.url != nil || result.text != nil) ? .mixed : .image
            } else if uti.contains("public.file-url") || uti.contains("public.data") {
                let mime = item["mimeType"] as? String ?? "application/octet-stream"
                let name = (item["suggestedName"] ?? item["filename"]) as? String ?? "shared-file"
                result.files.append(SharedFile(mimeType: mime, filename: name, size: dataStr.count))
                result.contentType = .file
            }
        }
        return result
    }

    private func normalizeAndroid(_ data: Any) -> NormalizedShareData {
        guard let dict = data as? [String: Any] else {
            return NormalizedShareData(platform: .android, contentType: .text)
        }

        var result = NormalizedShareData(platform: .android, contentType: .text)

        if let text = dict["EXTRA_TEXT"] as? String {
            if isUrl(text) {
                result.url = text
                result.contentType = .url
            } else {
                result.text = text
            }
        }
        if let subject = dict["EXTRA_SUBJECT"] as? String {
            result.title = subject
        }
        if result.text == nil, let html = dict["EXTRA_HTML_TEXT"] as? String {
            result.text = html
        }

        let mimeType = dict["type"] as? String ?? "application/octet-stream"
        if let streams = dict["EXTRA_STREAM"] {
            let streamList: [Any] = (streams as? [Any]) ?? [streams]
            for stream in streamList {
                if mimeType.hasPrefix("image/") {
                    result.images.append(SharedImage(mimeType: mimeType))
                    result.contentType = (result.text != nil || result.url != nil) ? .mixed : .image
                } else {
                    let name = (stream as? [String: Any])?["displayName"] as? String ?? "shared-file"
                    result.files.append(SharedFile(mimeType: mimeType, filename: name, size: 0))
                    result.contentType = .file
                }
            }
        }
        return result
    }

    private func normalizeWeb(_ data: Any) -> NormalizedShareData {
        guard let dict = data as? [String: Any] else {
            return NormalizedShareData(platform: .web, contentType: .text)
        }

        var result = NormalizedShareData(platform: .web, contentType: .text)

        if let t = dict["title"] as? String { result.title = t }
        if let t = dict["text"] as? String { result.text = t }
        if let u = dict["url"] as? String {
            result.url = u
            result.contentType = .url
        }

        // Web Share API Level 2: files
        if let files = dict["files"] as? [[String: Any]] {
            for file in files {
                let name = file["name"] as? String ?? "shared-file"
                let mime = file["type"] as? String ?? "application/octet-stream"
                let size = file["size"] as? Int ?? 0
                result.files.append(SharedFile(mimeType: mime, filename: name, size: size))
            }
            if !result.files.isEmpty {
                result.contentType = (result.url != nil || result.text != nil) ? .mixed : .file
            }
        }
        return result
    }

    private func normalizeGeneric(_ data: Any, platform: SharePlatform) -> NormalizedShareData {
        let dict = data as? [String: Any] ?? [:]
        return NormalizedShareData(
            text: dict["text"] as? String,
            url: dict["url"] as? String,
            title: dict["title"] as? String,
            platform: platform,
            contentType: .text
        )
    }
}

File: kits/data-integration/providers/capture_mode/share_intent.rs
// Data Integration Kit - Share Intent Capture Provider
// Mobile/OS share sheet receiver normalizing across iOS NSItemProvider and Android Intent extras

use std::collections::HashMap;

pub const PROVIDER_ID: &str = "share_intent";
pub const PLUGIN_TYPE: &str = "capture_mode";

#[derive(Debug, Clone)]
pub struct CaptureInput {
    pub url: Option<String>,
    pub file: Option<Vec<u8>>,
    pub email: Option<String>,
    pub share_data: Option<serde_json::Value>,
}

#[derive(Debug, Clone)]
pub struct CaptureConfig {
    pub mode: String,
    pub options: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone)]
pub struct SourceMetadata {
    pub title: String,
    pub url: Option<String>,
    pub captured_at: String,
    pub content_type: String,
    pub author: Option<String>,
    pub tags: Option<Vec<String>>,
    pub source: Option<String>,
}

#[derive(Debug, Clone)]
pub struct CaptureItem {
    pub content: String,
    pub source_metadata: SourceMetadata,
    pub raw_data: Option<String>,
}

#[derive(Debug)]
pub enum CaptureError {
    MissingShareData,
    ParseError(String),
}

impl std::fmt::Display for CaptureError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            CaptureError::MissingShareData => write!(f, "share_intent capture requires shareData"),
            CaptureError::ParseError(e) => write!(f, "Parse error: {}", e),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum SharePlatform {
    Ios,
    Android,
    Web,
    Desktop,
    Unknown,
}

impl SharePlatform {
    fn as_str(&self) -> &str {
        match self {
            SharePlatform::Ios => "ios",
            SharePlatform::Android => "android",
            SharePlatform::Web => "web",
            SharePlatform::Desktop => "desktop",
            SharePlatform::Unknown => "unknown",
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum ShareContentType {
    Text,
    Url,
    Image,
    File,
    Mixed,
}

impl ShareContentType {
    fn as_str(&self) -> &str {
        match self {
            ShareContentType::Text => "text",
            ShareContentType::Url => "url",
            ShareContentType::Image => "image",
            ShareContentType::File => "file",
            ShareContentType::Mixed => "mixed",
        }
    }

    fn mime_type(&self) -> &str {
        match self {
            ShareContentType::Image => "image/*",
            ShareContentType::Url => "text/uri-list",
            ShareContentType::File => "application/octet-stream",
            _ => "text/plain",
        }
    }
}

#[derive(Debug, Clone)]
pub struct SharedImage {
    pub mime_type: String,
    pub width: Option<u32>,
    pub height: Option<u32>,
}

#[derive(Debug, Clone)]
pub struct SharedFile {
    pub mime_type: String,
    pub filename: String,
    pub size: usize,
}

#[derive(Debug, Clone)]
pub struct NormalizedShareData {
    pub text: Option<String>,
    pub url: Option<String>,
    pub title: Option<String>,
    pub images: Vec<SharedImage>,
    pub files: Vec<SharedFile>,
    pub platform: SharePlatform,
    pub content_type: ShareContentType,
}

fn detect_platform(data: &serde_json::Value) -> SharePlatform {
    let obj = match data.as_object() {
        Some(o) => o,
        None => return SharePlatform::Unknown,
    };

    // iOS NSItemProvider pattern
    if obj.contains_key("itemProviders") || obj.contains_key("NSExtensionItem") || obj.contains_key("UTType") {
        return SharePlatform::Ios;
    }
    // Android Intent pattern
    if obj.contains_key("action") || obj.contains_key("EXTRA_TEXT") || obj.contains_key("EXTRA_STREAM") {
        return SharePlatform::Android;
    }
    // Web Share API
    if obj.contains_key("title") && (obj.contains_key("text") || obj.contains_key("url")) {
        if !obj.contains_key("action") && !obj.contains_key("itemProviders") {
            return SharePlatform::Web;
        }
    }
    // Desktop clipboard/drag-drop
    if obj.contains_key("clipboardData") || obj.contains_key("dataTransfer") {
        return SharePlatform::Desktop;
    }
    SharePlatform::Unknown
}

fn is_url(s: &str) -> bool {
    s.starts_with("http://") || s.starts_with("https://")
}

fn normalize_ios(data: &serde_json::Value) -> NormalizedShareData {
    let mut result = NormalizedShareData {
        text: None, url: None, title: None,
        images: Vec::new(), files: Vec::new(),
        platform: SharePlatform::Ios, content_type: ShareContentType::Text,
    };

    if let Some(ext) = data.get("NSExtensionItem").and_then(|v| v.as_object()) {
        if let Some(t) = ext.get("attributedTitle").and_then(|v| v.as_str()) {
            result.title = Some(t.to_string());
        }
    }

    let items = data.get("itemProviders").or_else(|| data.get("items"))
        .and_then(|v| v.as_array());

    if let Some(items) = items {
        for item in items {
            let uti = item.get("UTType").or_else(|| item.get("typeIdentifier"))
                .and_then(|v| v.as_str()).unwrap_or("");
            let data_str = item.get("data").and_then(|v| v.as_str()).unwrap_or("");

            if uti.contains("public.url") || uti.contains("public.plain-text") {
                if is_url(data_str) {
                    result.url = Some(data_str.to_string());
                    result.content_type = ShareContentType::Url;
                } else {
                    result.text = Some(data_str.to_string());
                }
            } else if uti.contains("public.image") {
                let mime = if uti.contains("png") { "image/png" } else { "image/jpeg" };
                result.images.push(SharedImage { mime_type: mime.to_string(), width: None, height: None });
                result.content_type = if result.url.is_some() || result.text.is_some() {
                    ShareContentType::Mixed
                } else { ShareContentType::Image };
            } else if uti.contains("public.file-url") || uti.contains("public.data") {
                let mime = item.get("mimeType").and_then(|v| v.as_str())
                    .unwrap_or("application/octet-stream");
                let name = item.get("suggestedName").or_else(|| item.get("filename"))
                    .and_then(|v| v.as_str()).unwrap_or("shared-file");
                result.files.push(SharedFile {
                    mime_type: mime.to_string(),
                    filename: name.to_string(),
                    size: data_str.len(),
                });
                result.content_type = ShareContentType::File;
            }
        }
    }
    result
}

fn normalize_android(data: &serde_json::Value) -> NormalizedShareData {
    let mut result = NormalizedShareData {
        text: None, url: None, title: None,
        images: Vec::new(), files: Vec::new(),
        platform: SharePlatform::Android, content_type: ShareContentType::Text,
    };

    if let Some(text) = data.get("EXTRA_TEXT").and_then(|v| v.as_str()) {
        if is_url(text) {
            result.url = Some(text.to_string());
            result.content_type = ShareContentType::Url;
        } else {
            result.text = Some(text.to_string());
        }
    }
    if let Some(subject) = data.get("EXTRA_SUBJECT").and_then(|v| v.as_str()) {
        result.title = Some(subject.to_string());
    }
    if result.text.is_none() {
        if let Some(html) = data.get("EXTRA_HTML_TEXT").and_then(|v| v.as_str()) {
            result.text = Some(html.to_string());
        }
    }

    let mime_type = data.get("type").and_then(|v| v.as_str()).unwrap_or("application/octet-stream");
    if let Some(streams) = data.get("EXTRA_STREAM") {
        let stream_list = if streams.is_array() {
            streams.as_array().unwrap().clone()
        } else {
            vec![streams.clone()]
        };

        for stream in &stream_list {
            if mime_type.starts_with("image/") {
                result.images.push(SharedImage { mime_type: mime_type.to_string(), width: None, height: None });
                result.content_type = if result.text.is_some() || result.url.is_some() {
                    ShareContentType::Mixed
                } else { ShareContentType::Image };
            } else {
                let name = stream.get("displayName").and_then(|v| v.as_str())
                    .unwrap_or("shared-file");
                result.files.push(SharedFile {
                    mime_type: mime_type.to_string(),
                    filename: name.to_string(),
                    size: 0,
                });
                result.content_type = ShareContentType::File;
            }
        }
    }
    result
}

fn normalize_web(data: &serde_json::Value) -> NormalizedShareData {
    let mut result = NormalizedShareData {
        text: None, url: None, title: None,
        images: Vec::new(), files: Vec::new(),
        platform: SharePlatform::Web, content_type: ShareContentType::Text,
    };

    if let Some(t) = data.get("title").and_then(|v| v.as_str()) { result.title = Some(t.to_string()); }
    if let Some(t) = data.get("text").and_then(|v| v.as_str()) { result.text = Some(t.to_string()); }
    if let Some(u) = data.get("url").and_then(|v| v.as_str()) {
        result.url = Some(u.to_string());
        result.content_type = ShareContentType::Url;
    }

    if let Some(files) = data.get("files").and_then(|v| v.as_array()) {
        for file in files {
            let name = file.get("name").and_then(|v| v.as_str()).unwrap_or("shared-file");
            let mime = file.get("type").and_then(|v| v.as_str()).unwrap_or("application/octet-stream");
            let size = file.get("size").and_then(|v| v.as_u64()).unwrap_or(0) as usize;
            result.files.push(SharedFile { mime_type: mime.to_string(), filename: name.to_string(), size });
        }
        if !result.files.is_empty() {
            result.content_type = if result.url.is_some() || result.text.is_some() {
                ShareContentType::Mixed
            } else { ShareContentType::File };
        }
    }
    result
}

fn normalize_share_data(share_data: &serde_json::Value) -> NormalizedShareData {
    let platform = detect_platform(share_data);
    match platform {
        SharePlatform::Ios => normalize_ios(share_data),
        SharePlatform::Android => normalize_android(share_data),
        SharePlatform::Web => normalize_web(share_data),
        _ => {
            let text = share_data.get("text").and_then(|v| v.as_str()).map(String::from);
            let url = share_data.get("url").and_then(|v| v.as_str()).map(String::from);
            let title = share_data.get("title").and_then(|v| v.as_str()).map(String::from);
            NormalizedShareData {
                text, url, title,
                images: Vec::new(), files: Vec::new(),
                platform, content_type: ShareContentType::Text,
            }
        }
    }
}

fn extract_hostname(url: &str) -> String {
    url.find("://")
        .and_then(|i| {
            let rest = &url[i + 3..];
            let end = rest.find('/').unwrap_or(rest.len());
            Some(rest[..end].to_string())
        })
        .unwrap_or_else(|| url.to_string())
}

pub struct ShareIntentCaptureProvider;

impl ShareIntentCaptureProvider {
    pub fn new() -> Self { Self }

    pub fn capture(&self, input: &CaptureInput, config: &CaptureConfig) -> Result<CaptureItem, CaptureError> {
        let share_data = input.share_data.as_ref().ok_or(CaptureError::MissingShareData)?;
        let normalized = normalize_share_data(share_data);

        let mut content_parts: Vec<String> = Vec::new();
        if let Some(ref title) = normalized.title { content_parts.push(format!("# {}", title)); }
        if let Some(ref url) = normalized.url { content_parts.push(format!("URL: {}", url)); }
        if let Some(ref text) = normalized.text { content_parts.push(text.clone()); }
        if !normalized.images.is_empty() {
            content_parts.push(format!("\nImages: {} shared", normalized.images.len()));
            for (i, img) in normalized.images.iter().enumerate() {
                let dim_info = img.width.map(|w| format!(" {}x{}", w, img.height.unwrap_or(0))).unwrap_or_default();
                content_parts.push(format!("  [{}] {}{}", i + 1, img.mime_type, dim_info));
            }
        }
        if !normalized.files.is_empty() {
            content_parts.push(format!("\nFiles: {} shared", normalized.files.len()));
            for (i, file) in normalized.files.iter().enumerate() {
                content_parts.push(format!("  [{}] {} ({}, {} bytes)", i + 1, file.filename, file.mime_type, file.size));
            }
        }

        let title = normalized.title.clone()
            .or_else(|| normalized.url.as_ref().map(|u| format!("Shared: {}", extract_hostname(u))))
            .unwrap_or_else(|| "Shared Content".to_string());

        let content = if content_parts.is_empty() {
            "(empty share)".to_string()
        } else {
            content_parts.join("\n")
        };

        let mut tags = vec![
            "share-intent".to_string(),
            normalized.platform.as_str().to_string(),
            normalized.content_type.as_str().to_string(),
        ];
        if !normalized.images.is_empty() { tags.push("has-images".to_string()); }
        if !normalized.files.is_empty() { tags.push("has-files".to_string()); }

        Ok(CaptureItem {
            content,
            source_metadata: SourceMetadata {
                title,
                url: normalized.url,
                captured_at: chrono::Utc::now().to_rfc3339(),
                content_type: normalized.content_type.mime_type().to_string(),
                author: None,
                tags: Some(tags),
                source: Some("share_intent".to_string()),
            },
            raw_data: None,
        })
    }

    pub fn supports(&self, input: &CaptureInput) -> bool {
        input.share_data.is_some()
    }
}
